.copilot/cpp_instructions.md

```diff
@@ -0,0 +1,88 @@
+
+#### 3\. Modern C++ Usage
+
+* **C++ Standard:** The codebase uses C++20 or newer. Prefer features from these standards, especially coroutines, as they improve code clarity and safety.
+* **Resource Management:** Use Seastar-specific smart pointers and resource management types.
+    * Prefer `foreign_ptr` for sharing resources between Seastar and other code.
+    * Use `lw_shared_ptr` and `shared_ptr` for shared ownership within the Seastar framework.
+    * Avoid `std::unique_ptr`, `std::shared_ptr`, and raw pointers unless there is a clear, documented reason (e.g., interfacing with C-style APIs).
+* **Headers:**
+    * Use `#pragma once` for include guards.
+    * Include project headers first, followed by third-party library headers, and finally standard library headers.
+* **No `using namespace std;`:** Never include `using namespace std;` in header files. In source files, use it sparingly and only within function scope to minimize naming conflicts.
+* **Avoid Macros:** Avoid preprocessor macros as much as possible. Use `inline` functions, `constexpr` variables, or templates instead.
+* **Type Safety:** Prefer scoped enums (`enum class`) over unscoped enums.
+* **`const` Correctness:** Mark all functions and variables as `const` whenever their state does not change. Use `const` references for function parameters to avoid unnecessary copies.
+
+# 4. Concurrency and Asynchrony
+
+* **Coroutines are Preferred:** The primary asynchronous programming model is coroutines. Prefer using coroutines over traditional callbacks or continuations for all new code.
+* **Seastar Framework:** Adhere to the concurrency model and best practices of the Seastar framework. The code should be written to take advantage of Seastar's shard-per-core architecture.
+* **Thread Safety:** Pay close attention to thread safety. Shared data should be protected by appropriate synchronization primitives (e.g., `seastar::mutex`, `std::atomic`), or structured to avoid sharing entirely (e.g., using message passing or per-core data).
+* **`co_await` Semantics:** When using `co_await`, ensure you understand the execution context and potential for context switching.
+* **Avoid Blocking:** Never perform blocking I/O or long-running computations on the reactor thread. Such operations must be offloaded to dedicated thread pools.
+
+# 5. Performance and Optimization
+
+* **Zero-Overhead Abstractions:** Prefer abstractions that have zero or minimal runtime overhead.
+* **Avoid Dynamic Allocation:** Minimize dynamic memory allocations, especially in hot code paths. Prefer stack-based or pre-allocated objects.
+* **Compiler Optimizations:** Write code that is easy for the compiler to optimize. Avoid aliasing, use `constexpr` where possible, and prefer standard library algorithms that can be vectorized.
+* **Benchmarking:** Any performance-critical change should be accompanied by benchmarks to validate its impact.
```

Comment:
Use fragmented containers (utils::chunked_vector) over std::vector where there is potential to break the 128k barrier.